modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericFunctionsTypeCoercionTest.java

```diff
@@ -0,0 +1,627 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.planner.datatypes;
+
+import static org.apache.ignite.internal.sql.engine.util.TypeUtils.native2relationalType;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexNode;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.NumericPair;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.TypePair;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.Types;
+import org.apache.ignite.internal.sql.engine.rel.IgniteTableScan;
+import org.apache.ignite.internal.sql.engine.schema.IgniteSchema;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.type.DecimalNativeType;
+import org.apache.ignite.internal.type.NativeType;
+import org.apache.ignite.internal.type.NativeTypes;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+import org.junit.jupiter.api.Named;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * A set of tests to verify behavior of type coercion for numeric functions, when operand belongs to the NUMERIC type family.
+ */
+public class NumericFunctionsTypeCoercionTest extends BaseTypeCoercionTest {
+
+    private static final List<NativeType> INT_TYPES = List.of(NativeTypes.INT8, NativeTypes.INT16, NativeTypes.INT32, NativeTypes.INT64);
+
+    @ParameterizedTest
+    @MethodSource("modArgs")
+    public void mod(
+            TypePair typePair,
+            Matcher<RexNode> arg1, Matcher<RexNode> arg2,
+            NativeType returnType
+    ) throws Exception {
+        NativeType type1 = typePair.first();
+        NativeType type2 = typePair.second();
+
+        IgniteSchema schema = createSchemaWithTwoColumnTable(type1, type2);
+
+        List<Matcher<RexNode>> args = List.of(arg1, arg2);
+        Matcher<RelNode> matcher = new FunctionCallMatcher(args).resultWillBe(returnType);
+
+        assertPlan("SELECT MOD(C1, C2) FROM T", schema, matcher::matches, List.of());
+    }
+
+    private static Stream<Arguments> modArgs() {
```

Comment:
first, please use NumericPair here. second, almost every test follows the same approach to define test cases:
```
forTypePair(NumericPair.TINYINT_DECIMAL_2_1)
                        .firstOpMatches(castTo(Types.DECIMAL_4_1))
                        .secondOpMatches(castTo(Types.DECIMAL_4_1)),
```
please follow this approach in new tests

Reply:
There is no INTEGER_TYPE1_INTEGER_TYPE2 type in `NumericPair`, 
It hasn't enough types to cover `MOD` function. What should I do about it?

Reply:
@korlov42 The form you mentioned precludes from checking a function return type. Should I remove checks of return types?

> forTypePair(NumericPair.TINYINT_DECIMAL_2_1)
>                        .firstOpMatches(castTo(Types.DECIMAL_4_1))
>                        .secondOpMatches(castTo(Types.DECIMAL_4_1)),

Reply:
Fixed.